### ISSUE TITLE
.NET should expose a generalized ASN.1 reader/writer API

### ISSUE THREAD

**Issue by bartonjs** [dotnet/runtime#22610]
A lot of features in cryptography and X.509 certificates rely on the ability to read or write ASN.1-structured data.  Not only would a good API for ASN.1 reading/writing simplify some of our existing code, it's very handy when someone wants to try adding support for something we don't have inbox, like reading a new certificate extension.

There are a couple of different aspects to the problem:

## 1) What do inputs/outputs look like?

One of the harder parts to remember to get right is rejecting unexpected data.  For example, from https://tools.ietf.org/html/rfc3280#section-4.1.1.2:

```asn
   AlgorithmIdentifier  ::=  SEQUENCE  {
        algorithm               OBJECT IDENTIFIER,
        parameters              ANY DEFINED BY algorithm OPTIONAL  }
```

The `AlgorithmIdentifier` structure defines two, and only two fields.  If a third field is present the data is invalid.  (In our current internal implementation this means you need `if (reader.HasData) throw new CryptographicException();` at the end of reading it).

This means that we really don't want to have to resort to a ReadNext() type API, since it's easy to do it poorly.

One strawman is to build on the model of existing serializers, so one would so something like

```C#
[StructLayout(StructLayoutKind.Sequential)]
internal struct AlgorithmIdentifier
{
    internal ObjectIdentifier Algorithm;
    [Optional]
    internal ??? Parameters;
}
```

The fact that the first example drew a need for a base class might be telling that it needs to be a Sequential-layout class instead of a struct.  Or maybe ASN-ANY has to be a byte[] that gets reinterpreted by the caller (but the serializer validates that it was legally encoded).

But sequential ordering on the type seems like a requirement.

C-based libraries generally seem to solve this problem with macro-based template languages.  So maybe a template is the way to go.  But, maybe not.

Needs to consider:

* OPTIONAL (DER says never to write down ASN-NULL for an OPTIONAL value)
* DEFAULT (DER says to never write down a default value)
  * Sometimes a default is on a complex type (like saying that a default AlgorithmIdentifier is SHA-2-256).
* ANY
* Context-specific tagging
  * And Application-specific, and Private.  I suppose
* That tags *technically* are not constrained to one-byte values.
* Collections (arrays/etc) can be SEQUENCE OF (write them down in the order I gave you) or SET OF (write them down in a proscribed order)
    * This *could* be skipped in the encoder and declared to be a problem for the data provider; but since the sort order for SET OF frequently comes down to the data length (before data values) it is sort of hard for them to do it right.
* ASN INTEGER is C# BigInteger.  But also supporting Int32/Int64/etc has virtue.
* There are lots of string encodings in ASN
* I'm sure floating point is unpleasant.
* DateTime values can be pretty weird.  X.509 adds more restrictions than DER does.  Is that an encoding?  An attribute?  A post-deserialization validation?
* Any other quirks that exist when combining ITU-T X.680 (ASN), ITU-T X.690 (BER/CER/DER) and C#.

## 2) What encodings are supported?

ITU-T X.690 defines 3 different encoding rulesets:

* BER (Basic Encoding Rules)
* CER (Canonical Encoding Rules)
* DER (Distinguished Encoding Rules)

Generally speaking;  BER defines all the possibilities for representing data, CER describes a set of restrictions on BER, and DER describes even more restrictions.  Occasionally CER and DER give conflicting recommendations, though.

* Read-as-BER can read anything.
* Read-as-CER (maybe) should reject an octet-string with a primitive encoding in excess of 1000 octets (ITU-T REC X.690 2015-08 section 9.2)
  * (Maybe there needs to be CER-strict and CER-flexible?)
* Read-as-DER should (definitely) reject any indefinite length encoding (ITU-T REC X.690 2015-08 section 10.1).

Almost all things in cryptography use DER, making DER a minimum bar.  But, since [CMS / S/MIME / PKCS#7](https://tools.ietf.org/html/rfc2315) says that it's designed to allow BER indefinite length encodings, there are clearly cases where readers need to be more flexible.  (Windows and OpenSSL both write all CMS objects with DER; macOS uses indefinite length encodings)

## 3) What does an API look like?

Largely this comes out of the inputs/outputs section, but encodings need to be taken into account.

```C#
public static AsnEncoding
{
    public static ???? Encode<T>(T value, AsnEncodingType type, Span<byte> output);
    public static ???? Decode<T>(ReadOnlySpan<byte> input, 
}
```

If there's a base class then there could be an OnDeserialize()-style validation callback per hydrated member.  If it's a convention that ends up being reflection, which isn't super fun.  So maybe it's just up to the user... but that's somewhat annoying for embedded structures.  Particularly through OPTIONAL.

-----

It probably makes sense to try to design the API, and use it internally, and then eventually expose it (essentially letting internal consumption be the proving ground).

**Comment by bartonjs** [comment]
There's now a spec/feature proposal at https://github.com/dotnet/designs/pull/93.

**Comment by bartonjs** [comment]
Fixed by #36729.
